src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+    let currentState = { ...state };
+    const stateHistory = [];
+
+    actions.forEach(action => {
+        switch (action.type) {
+            case 'clear':
+                currentState = {};
+                break;
+            case 'addProperties':
+                currentState = { ...currentState, ...action.extraData };
+                break;
+            case 'removeProperties':
+                action.keysToRemove.forEach(key => {
+                    delete currentState[key];
+                });
+                break;
+            default:
+                break;
+        }
+        stateHistory.push({ ...currentState });
```

Comment:
Pushing a copy of the object to the `stateHistory` array at the end of each loop cycle is correctly placed outside of the `switch` block. This adheres to the additional rule that it is better to do so after the current action is processed.